autopush/db.py

```diff
@@ -432,42 +444,60 @@ def generate_last_connect_values(date):
             yield int(val)
 
 
-def list_tables(client=g_client):
-    """Return a list of the names of all DynamoDB tables."""
-    start_table = None
-    while True:
-        if start_table:  # pragma nocover
-            result = client.list_tables(ExclusiveStartTableName=start_table)
-        else:
-            result = client.list_tables()
-        for table in result.get('TableNames', []):
-            yield table
-        start_table = result.get('LastEvaluatedTableName', None)
-        if not start_table:
-            break
+def table_exists(tablename, boto_resource=None):
+    # type: (str, DynamoDBResource) -> bool
+    """Determine if the specified Table exists"""
+    try:
+        return DynamoDBTable(boto_resource, tablename).table_status in [
+            'CREATING', 'UPDATING', 'ACTIVE']
+    except ClientError:
+        return False
 
 
-def table_exists(tablename, client=None):
-    """Determine if the specified Table exists"""
-    if not client:
-        client = g_client
-    return tablename in list_tables(client)
+class DynamoDBResource(threading.local):
+    def __init__(self, **kwargs):
+        conf = kwargs
+        # If there is no endpoint URL, we must delete the entry
+        if "endpoint_url" in conf and not conf["endpoint_url"]:
+            del(conf["endpoint_url"])
+        region = conf.get("region_name",
+                          os.getenv("AWS_REGION_NAME", "us-east-1"))
+        if "region_name" in conf:
+            del(conf["region_name"])
+        self._resource = boto3.resource(
+            "dynamodb",
+            config=botocore.config.Config(region_name=region),
+            **conf)
+
+    def __getattr__(self, name):
+        return getattr(self._resource, name)
+
+
+class DynamoDBTable(threading.local):
+    def __init__(self, ddb_resource, *args, **kwargs):
+        # type: (DynamoDBResource, *Any, **Any) -> None
+        self._table = ddb_resource.Table(*args, **kwargs)
+
+    def __getattr__(self, name):
+        return getattr(self._table, name)
 
 
 class Message(object):
     """Create a Message table abstraction on top of a DynamoDB Table object"""
-    def __init__(self, table, metrics, max_ttl=MAX_EXPIRY):
-        # type: (Table, IMetrics) -> None
+    def __init__(self, table, metrics=None, boto_resource=None,
+                 max_ttl=MAX_EXPIRY):
+        # type: (DynamoDBTable, IMetrics, DynamoDBResource, int) -> None
```

Comment:
Why does it need the boto_resource?

Reply:
mostly for consistency and testing purposes. `DynamoDBTable` still needs to pull from the resource, so having the explicit resource passed ensures that nothing "magical" happens.